### ISSUE TITLE
tablets: support dc/rack change

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#18024]
Follow-up https://github.com/scylladb/scylladb/issues/17513
based on Benny's comment https://github.com/scylladb/scylladb/issues/17513#issuecomment-2018920662

> With vnodes, if nodes dc changes due to snitch change, running repair/rebuild should populate the now missing token replucas based on the token rung, and running cleanup will delete now-disowned tokens.

> With tablets that ain't going to work and we need a way to migrate tablets in order to bring the replicas back into agreement with the rf configuration.

**Comment by bhalevy** [comment]
[`test/topology_custom/test_snitch_change.py`](https://github.com/scylladb/scylladb/blob/master/test/topology_custom/test_snitch_change.py) was added for this case.
We should test it also with tablets.
Based on analysis (See https://github.com/scylladb/scylladb/issues/17513#issuecomment-2021611547), the expectation is that we'll have a problem with consistency_level = EACH_QUORUM.

There is no way this could be addressed for 6.0 in the current timeframe,
so we should document the limitations with tablets.
Cc @annastuchlik

**Comment by bhalevy** [comment]
@paszkow / @pwrobelse can you please look into this issue?
Let's start with reusing and extending `test/topology_custom/test_snitch_change.py` for tablets.
we should also add read and write queries after the dc renaming with all supported consistency levels to map if and where exactly we hit issues.
(It will be good to do that also with consistent-topology-changes and vnodes)

**Comment by tgrabiec** [comment]
> [`test/topology_custom/test_snitch_change.py`](https://github.com/scylladb/scylladb/blob/master/test/topology_custom/test_snitch_change.py) was added for this case.
> We should test it also with tablets.

The test case is a full DC rename, so it should work with tablets.

> Based on analysis (See [#17513 (comment)](https://github.com/scylladb/scylladb/issues/17513#issuecomment-2021611547)), the expectation is that we'll have a problem with consistency_level = EACH_QUORUM.

But vnodes would suffer from the same problem with EACH_QUORUM in that scenario, that is until DC is renamed in keyspace definition, the old DC will appear as unavailable.

**Comment by dani-tweig** [comment]
@tzach 
We need your input here.
What is the scope, priority and is this use case common enough?

**Comment by dani-tweig** [comment]
@tzach ^

**Comment by tzach** [comment]
> What is the scope, priority and is this use case common enough?

Adding and removing DCs and Rack is a common use case and must be supported ASAP.

**Comment by dani-tweig** [comment]
> Adding and removing DCs and Rack is a common use case and must be supported ASAP.

I see it was marked as Q4 content.

**Comment by bhalevy** [comment]
> > What is the scope, priority and is this use case common enough?
> 
> Adding and removing DCs and Rack is a common use case and must be supported ASAP.

@tzach but this issue is not about adding or removing DCs or RACKs.
It's about renaming them via the snitch.

Note: using the snitch to "move" a node between RACKs or DCs should be banned.
It cannot work with NetworkTopologyStrategy.  One better decommission a node, wipe it clean, and restart it with the new DC/RACK settings.

**Comment by bhalevy** [comment]
Ref https://github.com/scylladb/scylladb/issues/12306
Cc @xemul

**Comment by bhalevy** [comment]
I suggest we start with laying out an offline solution, that requires a full cluster restart, so that the snitch can be reconfigured when the nodes are down. And then, hopefully one only needs to change the NTS replication factor settings after the nodes are restarted to "move" the replication factor from the old dc name to the new dc name (not needed when only the rack is renamed).

When this is working and tested we can try to think of an online procedure requiring only rolling-restart.
Though it may cause transient Unavailable errors as nodes will disappear from one dc/rack and reappear in a different, new one.

**Comment by mykaul** [comment]
I'm not sure why it's a P2 and not a P4. Seems like a very low priority item to me - perhaps I'm missing a use case.
If we put any efforts into such items, let's remove the snitches from the code into standalone scripts (that can be re-used by the drivers, etc. - there's an issue for it somewhere...)

**Comment by vladzcloudius** [comment]
I don't think DC and/or RACK rename of an existing node should be ever allowed.
On the contrary - this should be strictly forbidden because this is going to change the replication without streaming == data loss.

**Comment by bhalevy** [comment]
See https://github.com/scylladb/scylladb/issues/22869#issuecomment-2664747823 for a hand-waved proposal to manage the DCs and RACKs in a system table over raft, and use UUIDs for them internally so we can manage the DC and RACK name in the system table (to be referred by name over CQL).

Nodes GossipingPropertyFileSnitch will be considered only once when the node bootstraps, and once it joins the cluster its DC or RACK can't be changed via the snitch.  To move the node between RACKs or DCs one has to decommission or replace it first and then bootstrap it in the new location as a new node.

We can then offer APIs to rename existing DCs or RACKs in the system table without affecting the nodes logical location in topology and the replication options in the schema since both would use the internal DC/RACK UUIDs that won't change by rename.

Cc @tgrabiec

**Comment by vladzcloudius** [comment]
> See [#22869 (comment)](https://github.com/scylladb/scylladb/issues/22869#issuecomment-2664747823) for a hand-waved proposal to manage the DCs and RACKs in a system table over raft, and use UUIDs for them internally so we can manage the DC and RACK name in the system table (to be referred by name over CQL).
> 
> Nodes GossipingPropertyFileSnitch will be considered only once when the node bootstraps, and once it joins the cluster its DC or RACK can't be changed via the snitch. To move the node between RACKs or DCs one has to decommission or replace it first and then bootstrap it in the new location as a new node.
> 
> We can then offer APIs to rename existing DCs or RACKs in the system table without affecting the nodes logical location in topology and the replication options in the schema since both would use the internal DC/RACK UUIDs that won't change by rename.
> 
> Cc [@tgrabiec](https://github.com/tgrabiec)

That's of course a nice idea. We thought about pretty much the same idea in CX independently.
I think we are on the same page here which is: the moment node bootstrapped it can't be moved to a different DC and/or RACK.

**Comment by bhalevy** [comment]
> > See [#22869 (comment)](https://github.com/scylladb/scylladb/issues/22869#issuecomment-2664747823) for a hand-waved proposal to manage the DCs and RACKs in a system table over raft, and use UUIDs for them internally so we can manage the DC and RACK name in the system table (to be referred by name over CQL).
> > Nodes GossipingPropertyFileSnitch will be considered only once when the node bootstraps, and once it joins the cluster its DC or RACK can't be changed via the snitch. To move the node between RACKs or DCs one has to decommission or replace it first and then bootstrap it in the new location as a new node.
> > We can then offer APIs to rename existing DCs or RACKs in the system table without affecting the nodes logical location in topology and the replication options in the schema since both would use the internal DC/RACK UUIDs that won't change by rename.
> > Cc [@tgrabiec](https://github.com/tgrabiec)
> 
> That's of course a nice idea. We thought about pretty much the same idea in CX independently.

Great!

> I think we are on the same page here which is: the moment node bootstrapped it can't be moved to a different DC and/or RACK.

Yes. We're in agreement on that.

**Comment by bhalevy** [comment]
@emdroid reassigning to you as this is essentially part of raft topology.
That said, I'm not sue if and how we'll ever get to it.
Right now, first priority is to focus on safety and prevent changes in DC/RACK after bootstrap (#22869).

**Comment by dani-tweig** [comment]
Closing this issue as it was moved to Jira.
Please continue the thread in https://scylladb.atlassian.net/browse/SCYLLADB-1967
